python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,198 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from abc import ABC, abstractmethod
+from typing import Optional
+
+from iceberg.schema import Schema
+from iceberg.table.base import PartitionSpec, Table
+
+
+class Catalog(ABC):
+    """
+    Base Catalog for table operations like - create, drop, load, list and others.
+
+    Attributes:
+        name(str): Name of the catalog
+        properties(dict): Catalog properties
+    """
+
+    def __init__(self, name: str, properties: dict):
+        self._name = name
+        self._properties = properties
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def properties(self) -> dict:
+        return self._properties
+
+    @abstractmethod
+    def list_tables(self) -> list:
+        """
+        List tables in the catalog.
+
+        :return: list of table names in the catalog.
+        """
+
+    @abstractmethod
+    def create_table(
+        self,
+        name: str,
+        schema: Schema,
+        partition_spec: PartitionSpec,
```

Comment:
This should be optional to account for unpartitioned table creation, right? Actually thinking about this more, information like name, schema, and partition spec should be contained in an instance of `Table`--would it be better here if these table related methods just took a `Table` instance?

Reply:
Agreed, they should ideally be taking a `TableBuilder` instance which has the table spec and the method returns  the created 'Table' instance. In my mental model, `TableBuilder` is a spec that has all the information required to create a 'Table', and 'Table' is something that holds the existing table state. Are you thinking on similar lines as well?

Reply:
In Java, we use a dedicated Unpartitioned spec.

But I agree that this should be optional and then the builder (or this function… either way) should default to Unpartitioned if this is None (implicitly using the Unpartitioned spec).

Reply:
Added a frozen `TableSpec` class in a recent commit to hold user provided table specifications to create or replace a table. Also, I made partition_spec optional in there.

Reply:
I've removed the `TableSpec` wrapper class in a recent commit and reverted back to taking in exploded keyword arguments. I've made the `partition_spec` optional in the signature now.